libsolidity/lsp/RenameSymbol.cpp

```diff
@@ -0,0 +1,321 @@
+using namespace solidity::langutil;
+using namespace solidity::lsp;
+using namespace std;
+
+namespace
+{
+
+CallableDeclaration const* extractCallableDeclaration(FunctionCall const& _functionCall)
+{
+	if (
+		auto const* functionType = dynamic_cast<FunctionType const*>(_functionCall.expression().annotation().type);
+		functionType && functionType->hasDeclaration()
+	)
+		if (auto const* functionDefinition = dynamic_cast<FunctionDefinition const*>(&functionType->declaration()))
+			return functionDefinition;
+
+	return nullptr;
+}
+
+}
+
+void RenameSymbol::operator()(MessageID _id, Json::Value const& _args)
+{
+	auto const&& [sourceUnitName, lineColumn] = extractSourceUnitNameAndLineColumn(_args);
+	string const newName = _args["newName"].asString();
+	string const uri = _args["textDocument"]["uri"].asString();
+
+	ASTNode const* sourceNode = m_server.astNodeAtSourceLocation(sourceUnitName, lineColumn);
+
+	m_symbolName = {};
+	m_declarationToRename = nullptr;
+	m_sourceUnits = { &m_server.compilerStack().ast(sourceUnitName) };
+	m_locations.clear();
+
+	optional<int> cursorBytePosition = charStreamProvider()
+		.charStream(sourceUnitName)
+		.translateLineColumnToPosition(lineColumn);
+	solAssert(cursorBytePosition.has_value(), "Expected source pos");
+
+	extractNameAndDeclaration(*sourceNode, *cursorBytePosition);
+
+	// Find all source units using this symbol
+	for (auto const& [name, content]: fileRepository().sourceUnits())
+	{
+		auto const& sourceUnit = m_server.compilerStack().ast(name);
+		for (auto const* referencedSourceUnit: sourceUnit.referencedSourceUnits(true, util::convertContainer<set<SourceUnit const*>>(m_sourceUnits)))
+			if (*referencedSourceUnit->location().sourceName == sourceUnitName)
+			{
+				m_sourceUnits.insert(&sourceUnit);
+				break;
+			}
+	}
+
+	// Origin source unit should always be checked
+	m_sourceUnits.insert(&m_declarationToRename->sourceUnit());
+
+	Visitor visitor(*this);
+
+	for (auto const* sourceUnit: m_sourceUnits)
+		sourceUnit->accept(visitor);
+
+	// Apply changes in reverse order (will iterate in reverse)
+	sort(m_locations.begin(), m_locations.end());
+
+	Json::Value reply = Json::objectValue;
+	reply["changes"] = Json::objectValue;
+
+	Json::Value edits = Json::arrayValue;
+
+	for (auto i = m_locations.rbegin(); i != m_locations.rend(); i++)
+	{
+		solAssert(i->isValid());
+
+		// Replace in our file repository
+		string const uri = fileRepository().sourceUnitNameToUri(*i->sourceName);
+		string buffer = fileRepository().sourceUnits().at(*i->sourceName);
+		buffer.replace((size_t)i->start, (size_t)(i->end - i->start), newName);
+		fileRepository().setSourceByUri(uri, std::move(buffer));
+
+		Json::Value edit = Json::objectValue;
+		edit["range"] = toRange(*i);
+		edit["newText"] = newName;
+
+		// Record changes for the client
+		edits.append(edit);
+		if (i + 1 == m_locations.rend() || (i + 1)->sourceName != i->sourceName)
+		{
+			reply["changes"][uri] = edits;
+
```

Comment:
Could this blank line be removed? It seems unnecessary to me.

Reply:
I had it simply so that the body is not all sticking together and easier to read
Removed it now.

Reply:
I imagined that, but only two lines should be ok :)